multisig-contract/contracts/MultiSig.aes

```diff
@@ -0,0 +1,74 @@
+// ISC License
+//
+// Copyright (c) 2017, aeternity developers
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
+// OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+
+// THIS IS NOT SECURITY AUDITED
+// DO NEVER USE THIS WITHOUT SECURITY AUDIT FIRST
+
+contract MultiSig =
+
+  record state =
+    { admin_one : address
+    , admin_two : address
+    , multi_sig_hashes: map(address, hash) }
+
+  record spend_data =
+    { spend_to_address : address
+    , spend_amount : int }
+
+  public function init(admin_one' : address, admin_two' : address) : state =
+    require_not_equal_addresses(admin_one', admin_two')
+
+    { admin_one = admin_one',
+      admin_two = admin_two',
+      multi_sig_hashes = {}}
+
+  public stateful function add_data_and_spend_if_both_admins_agree(spend_data : spend_data) : bool =
+    require_caller_is_either_admin(Call.caller)
+
+    // update agreement to spend, so always latest agreement is in state for each admin
+    let data_hash = Crypto.blake2b(spend_data)
+    put(state{ multi_sig_hashes[Call.caller] = data_hash })
+
+    // if both admins agree on the same spend data, execute spend and reset agreement
+    let hash_admin_one_entry = Map.lookup(state.admin_one, state.multi_sig_hashes)
+    let hash_admin_two_entry = Map.lookup(state.admin_two, state.multi_sig_hashes)
+
+    switch((hash_admin_one_entry, hash_admin_two_entry))
+      (Some(hash_admin_one), Some(hash_admin_two)) =>
+        if(hash_admin_one == hash_admin_two)
+          put(state{ multi_sig_hashes = {} })
+          Chain.spend(spend_data.spend_to_address, spend_data.spend_amount)
```

Comment:
If this fails, we have a `revert` of the whole transaction so no need for checking the return value, right?

Reply:
Also `spend` returns a unit value, so there's nothing to check.